src/hotspot/os/posix/signals_posix.cpp

```diff
@@ -1339,6 +1333,21 @@ void install_signal_handlers() {
     (*end_signal_setting)();
   }
 
+  if (!ReduceSignalUsage) {
+    // This is just for early initialization phase. Intercepting the signal here
+    // reduces the risk that an attach client accidentally forces HotSpot to quit
+    // prematurely. Late initialization will overwrite BREAK_SIGNAL's handler to
+    // UserHandler.
+    // Note that HotSpot does NOT support signal chaining for BREAK_SIGNAL.
+    // We have to set it outside the window bounded by libjsig's
+    // JVM_begin_signal_setting and JVM_end_signal_setting above, because the
+    // window is intended for signals that support chaining. Otherwise libjsig
+    // would prevent us from overwriting BREAK_SIGNAL's handler to UserHandler.
```

Comment:
Okay but doesn't that also mean that libjsig will now allow application code to install its own BREAK_SIGNAL handler, when it should be prevented from doing so???

Reply:
Yes, this seems always the case regardless of this change. The application should use `-Xrs`/`-XX:+ReduceSignalUsage` if it needs to install its own handler for BREAK_SIGNAL.

This is also the reason why TestSigquit.java sets `-Xrs`, see test/hotspot/jtreg/runtime/signal/README:
> SIGQUIT, SIGTERM, SIGINT, and SIGHUP signals cannot be chained.
> If the application needs to handle these signals, the -Xrs option needs
> to be specified. So, test these signals only with -Xrs flag.

Reply:
My point is that this change seems to allow installation of a BREAK_SIGNAL handler even without -Xrs!

Reply:
> My point is that this change seems to allow installation of a BREAK_SIGNAL handler even without -Xrs!

Got it. This is indeed the case. However, prior to [JDK-8279124](https://bugs.openjdk.org/browse/JDK-8279124), HotSpot also behaved in the same way, that it allows installation of a custom BREAK_SIGNAL handler even without -Xrs.

I verified with a modified TestSigquit.java, [here](https://github.com/caoman/jdk/commit/346132aa5af8d59406d0b98567d18d16868a3a32). It tests the "postpre" and "postpost" mode without -Xrs, which both uses libjsig.so:
- Without any change, the test times out waiting for the custom signal handler to be called.
- With JDK-8279124 [backed out](https://github.com/caoman/jdk/commit/0983baad17d78cf64e1f604abb232e581700a662), OR with change in this PR, the test passes.

Since HotSpot does not support chaining for SIGQUIT, I think it is OK to leave the behavior undefined if user installs a custom SIGQUIT handler without -Xrs. I.e. the custom handler may or may not work, and it could break HotSpot functionality. libjsig is not designed to prevent installing custom SIGQUIT handler.

Reply:
I am somewhat surprised by this. My reading of the jsig code is that once the VM has installed a SIGQUIT handler during VM init, then subsequent attempts to install a user handler will add the handler to the set to be chained, but because this is a VM signal the VM will never call that chained handler. With the proposed changes if the SIGQUIT handler has not been installed by the VM then the user handler will be installed directly - though I think the VM's handler would then replace that.

I find this all rather hard to follow and evaluate. As this seems to fix an issue I guess we can take it as-is and then keep an eye out for any unintended consequences. Though they will likely only come from external users so not something we will know till after the next release.

Reply:
Updated with tstuefe@'s recommendations. It still keeps the behavior that a user can install a SIGQUIT handler without -Xrs.

Reply:
The behavior has changed. Now that with libjsig, user must have -Xrs to install a custom SIGQUIT handler.